### ISSUE TITLE
[MNG-5102] Mixin POM fragments

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6814]
**[Anthony Whitford](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=awhitford)** opened **[MNG-5102](https://issues.apache.org/jira/browse/MNG-5102?redirect=false)** and commented

I am looking for a way to _mixin_ POM fragments into POMs.  Note that this idea is beyond parent pom inheritance because all projects inherit from a corporate parent pom.  The problem that I am running into is that the corporate parent pom is turning into an _"everything but the kitchen sink"_ POM and I'd like to dissect it into POM fragments relevant for individual modules.

For example, I would like to have mixins for:
* Java projects (that include static code analysis plugins, javadoc, etc.)
* JPA projects (that include DDL generation)
* Flex projects (that include flexmojos, asdoc, etc.)
* Scala projects (that include the maven-scala-compiler plugin, scaladoc, etc.)
* JavaScript projects (that include build plugins like maven-yuicompressor-plugin with jslint and compress goals)

Hopefully, you get the idea.  Without the ability to factor pom logic, we are left with two symptoms:
1. copy/paste duplication
2. complex _"it does it all"_ parent poms (which slow down builds because more plugins are loaded even though they might not do anything material)

Note that a project may include multiple mixins as I could have a project that contains Java code, Scala code, and JavaScript.

Another idea is that the mixins could be parameterized, so that the ultimate pom can be customized based on the parameters (like tokens).

I recall reading about Mixins coming in Maven 3.1, but could not find any such issue to watch, so am creating one.

---

**Affects:** 2.2.1

**Attachments:**
- [daddy3.zip](https://issues.apache.org/jira/secure/attachment/12712704/daddy3.zip) (_5.47 kB_)
- [maven-tiles.zip](https://issues.apache.org/jira/secure/attachment/12713095/maven-tiles.zip) (_47.48 kB_)

**Issue Links:**
- [MNG-7844](https://issues.apache.org/jira/browse/MNG-7844) Support for DTD entities and XInclude in the code at build time

- [MNG-5588](https://issues.apache.org/jira/browse/MNG-5588) Scope import in pluginManagement

- [MNG-5173](https://issues.apache.org/jira/browse/MNG-5173) Include other XML files from POM

**Remote Links:**
- [Page
](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=65875544)
- [GitHub Pull Request #1209
](https://github.com/apache/maven/pull/1209)

84 votes, 95 watchers

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

yes, surprisingly there doesn't appear to be an issue for it, though lots of related ones. Adding to that version.

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

I've been in greater need of this lately - I'll kick a discussion on the development list

**Comment by jira-importer** [comment]
**[maoo](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maoo)** commented

Hi,
I've some code to share - tested on Maven 3.0.2 - which composes a Maven POM from different tiles; it's a spike of 2 days work, but I think results are very interesting.

The implementation consists of:
- an implementation of MavenLifecycleParticipant.java (120 lines)
- a components.xml

The attached ZIP file contains a Maven build to produce a JAR that needs to be copied in M2_HOME/lib, a README.txt and a working sample; there are still some issues (e.g. properties inheritance), which are reported in the README, but I think that it gives the idea.

I really hope this could help with this feature implementation.

**Comment by jira-importer** [comment]
**[Tony Lampada](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tonylampada)** commented

Maurizio, your maven-tiles seems very promising.

Do you still use it today?
Have you made any improvements to it?
By using it, have you bumped into new problems that you haven't thought of?
How about making it into a github project, or something?

**Comment by jira-importer** [comment]
**[Paul Benedict](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ID134349)** commented

Use "import" scope; that's how you can define a POM for particular types of projects (see issue description) and then add those dependencies to your project. I would say the "import" scope is the Mixin you want.

http://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html

**Comment by jira-importer** [comment]
**[Christoph Henrici](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chenrici)** commented

Unfortunately  does e2Eclipse not work with https://github.com/repaint-io/maven-tiles. It simply ignores it.
Exactly http://books.sonatype.com/mvnref-book/reference/pom-relationships-sect-pom-best-practice.html is simply not enough. Dependency Management is almost more urgent.

**Comment by jira-importer** [comment]
**[Steven Schlansker](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stevenschlansker)** commented

I ran into this again today -- we are going to end up copy and pasting hundreds of line of \<build> configuration between projects because we can't mix it in :(

I know this is only wishlist priority, but for me personally and probably the community at large, this would be a groundbreaking new feature!

**Comment by jira-importer** [comment]
**[Fabrizio Lungo](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=flungo)** commented

This could be done with profiles if MNG-6000 were implemented although it might still be nice to have a way of including poms to stop a parent pom that may contain many profiles becoming thousands of lines long.

**Comment by jira-importer** [comment]
**[Sam Brougher](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sbrougher)** commented

The inability to have some kind of mixin functionality is likely going to drive my company away from Maven to something like Gradle. This should really be a much higher priority than "maybe someday".

**Comment by jira-importer** [comment]
**[Nick Groenewegen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=nickgr6)** commented

+1

Was referenced to this issue by a StackOverflow comment (https://stackoverflow.com/q/48154136/639033).

We are also very much in need of this functionality. Our parent pom doesn't fit our need anymore.

Has this ticket any priority or should I look at something else?

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Can you please post on the users list what exactly does not work etc.

**Comment by jira-importer** [comment]
**[Mirko Raner](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mirko@raner.ws)** commented

+1

This functionality would be extremely useful to reduce code duplication in POMs.

My particular use case is the configuration of several executions for a plugin, all of which are 95% identical except one line, which is different for each execution. Right now, I have to duplicate the common part of the configuration several times in the POM.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

[Mirko Raner](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mirko@raner.ws) If you have parts of the configuration in common for the same plugin you can use things like this:

```xml
 <plugin>
   <..>
   <configuration>
    <The Common Configuration identical for all executions>
   </configuration>
   <executions>
     <execution>
       <id>first</id>
       <configuration>
         <The Special Line>
       </configuration>
        ..
     </execution>
     <execution>
       <id>second</id>
       <configuration>
         <The Special Line>
       </configuration>
        ..
     </execution>
   </executions>
</plugin>
```

**Comment by jira-importer** [comment]
**[Gordon Pettey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gpettey)** commented

Would vote a dozen times on this (and MNG-5588) if I could... parent pom here is 1280 lines and getting ever larger to accommodate plugin configurations that are needed by some but not all child poms. Using middle-level parent poms is not feasible as the plugin sets are neither a union nor entirely disparate. Would need so many middle-level parents the naming scheme alone would be a problem.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

This requires a change in the pom definition, but as long as there is no solution to port it back to model 4.0.0 once deployed, we can't implement it.
In the meantime https://github.com/repaint-io/maven-tiles should be able to help (no experience with it myself)

**Comment by jira-importer** [comment]
**[Andrew VanLoo](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gumshoe2055)** commented

Is it possible to use some kind of `include`s directive, like Apache Web Server does with configuration files?  That works extremely well there.

I just found: https://issues.apache.org/jira/browse/MNG-5173

Although, I am not altogether clear on why abritrary POMs will break things...

**Comment by jira-importer** [comment]
**[Maximilian Schulz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulzm)** commented

Is there any timeline when this might be implemented? Not having the possibility for composing pom's results in a lot of copy pasta ...

**Comment by jira-importer** [comment]
**[Delany](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=delany)** commented

Why don't we just use [Xinclude](https://www.w3.org/TR/2016/NOTE-xinclude-11-20160721)

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd" xmlns:xi="http://www.w3.org/2001/XInclude">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.turvy.topsy</groupId>
  <artifactId>thingamabob</artifactId>
  <version>1.0.0</version>
  <xi:include href="modules.xml"/>
  <xi:include href="dependencies.xml"/>
  <xi:include href="build.xml"/>
  <xi:include href="profiles.xml"/>
</project>
```

**Comment by jira-importer** [comment]
**[Romain Manni-Bucau](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=romain.manni-bucau)** commented

+1 for [Delany](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=delany)  idea

**Comment by jira-importer** [comment]
**[Jimisola Laursen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER281352)** commented

This issue is 10+ years old and unassigned and I'm therefore wondering if itactual consideration for 4.x ([Version Issues to be reviewed for 4.x) or not until 5.x due to the required change in POM Model?

This is probably the most crucial feature (to remove a major pain point) that our team is hoping for since it would allow us to break up our ever-growing parent POM. It would also allow us to collaborate easier with other organizations.

**Comment by jira-importer** [comment]
**[Jimisola Laursen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER281352)** commented

The [Delany](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=delany) idea using Xinclude does not allow for usingstandard  Maven versioning when using mixins. We also want to use Maven Repository.

Compare with the [proposal in Maven POM v5](https://cwiki.apache.org/confluence/display/MAVEN/POM+Model+Version+5.0.0#POMModelVersion5.0.0-%3Cmixin%3Eelement)

**Comment by jira-importer** [comment]
**[Jimisola Laursen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jimisola)** commented

FYI, our team opted to go with [Maven Tiles](https://github.com/repaint-io/maven-tiles).
Although it has worked well for us there is, of course, a slight fair that Maven Tiles will break or a bug will come up that is not fixed. Since, it's crucial for our production we'd much rather see something included as standard functionality in Maven.

**Comment by jira-importer** [comment]
**[Jimisola Laursen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jimisola)** commented

[gnodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnodet)  Been getting notification on your commits towards this issue/feature.

I'm curious, what is your plan for this issue with your commits?

**Comment by jira-importer** [comment]
**[Guillaume Nodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnt)** commented

I'd like to focus on bug fixing to get 4.0 out asap, but mixins can definitely be planned for 4.1.

**Comment by jira-importer** [comment]
**[Mark Derricutt](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=talios)** commented

[Jimisola Laursen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jimisola) I've been doing some experimentation with [gnodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnodet) 's PR and reusing my tiles - as I'm not only invested this as the Tiles Maven Plugin maintainer, but also user (and knowing Tiles doesn't work with 4 - too many deprecated APIs, some internal class abuse, and M4 model changes make it less than ideal ) - so far things looks like good for simple tiles (aka non-nested tiles).

Posted about this both on the PR and LinkedIn at https://www.linkedin.com/posts/mark-derricutt-46277_tiles-maven-apache-activity-7315306110662426624-zpkp

There's already A LOT of big changes coming in 4, so the sooner that can get finalised and shipped the better IMHO. Whilst tiles is getting more and more adoption, it's still fairly niche in the maven ecosystem to be blocker.

As I said in my post, at least gives a planned way forward.

 

[gnodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnodet) One thing I like about tiles, which I immediately missed with the mixin PR was some form of debuggability over what mixins are being applied, altho I've not yet ran with `-X` to see if any debug information is mentioned there.  We also rewrite execution IDs with the mixing GAV - I may have missed it over the years, but is there way in the effective-pom to see WHERE plugins etc. may come from (either mixin or parent).

**Comment by Frettman** [comment]
Will Mixins (at least in their current PR form) support extensions?
